client/src/components/JurisdictionAdmin/BatchRoundDataEntry.tsx

```diff
@@ -0,0 +1,645 @@
+  justify-content: space-between;
+  margin-top: 16px;
+
+  .${Classes.INPUT}.${HIDDEN_INPUT_CLASS_NAME} {
+    display: none;
+  }
+`
+
+const SheetActions = styled('div')`
+  align-items: center;
+  display: flex;
+`
+
+const ConfirmSheetRenameButton = styled(Button)`
+  margin-left: 8px;
+`
+
+const SheetRenameErrorMessage = styled('span').attrs({
+  className: Classes.TEXT_SMALL,
+})`
+  color: ${Colors.RED2};
+  margin-left: 8px;
+`
+
+const Spacer = styled('div')`
+  flex-grow: 1;
+`
+
+interface IBatchRoundDataEntryFormState {
+  resultTallySheets: IBatchResultTallySheet[]
+}
+
+const VOTE_TOTALS_TAB_ID = 'vote-totals'
+
+function defaultResultTallySheetName(index: number) {
+  return `Sheet ${index + 1}`
+}
+
+function constructEmptyResultTallySheet(index: number): IBatchResultTallySheet {
+  return {
+    name: defaultResultTallySheetName(index),
+    results: {},
+  }
+}
+
+interface IProps {
+  electionId: string
+  jurisdictionId: string
+  roundId: string
+}
+
+const BatchRoundDataEntry: React.FC<IProps> = ({
+  electionId,
+  jurisdictionId,
+  roundId,
+}) => {
+  const batchesResp = useBatches(electionId, jurisdictionId, roundId)
+  const contests = useContestsJurisdictionAdmin(electionId, jurisdictionId)
+  const recordBatchResults = useRecordBatchResults(
+    electionId,
+    jurisdictionId,
+    roundId
+  )
+  const { confirm, confirmProps } = useConfirm()
+
+  const [searchQuery, setSearchQuery] = useState('')
+  const [selectedBatchId, setSelectedBatchId] = useState<IBatch['id']>()
+  const [selectedTabId, setSelectedTabId] = useState('')
+  const [isEditing, setIsEditing] = useState(false)
+  const [isRenamingSheet, setIsRenamingSheet] = useState(false)
+  const [sheetNameBeforeRename, setSheetNameBeforeRename] = useState<{
+    index: number
+    name: string
+  }>()
+
+  const batches = batchesResp.isSuccess ? batchesResp.data.batches : []
+  const filteredBatches = batches.filter(batch =>
+    batch.name.toLowerCase().includes(searchQuery.toLowerCase())
+  )
+  const selectedBatch = batches.find(batch => batch.id === selectedBatchId)
+
+  const initialResultTallySheets = selectedBatch?.resultTallySheets.length
+    ? selectedBatch.resultTallySheets
+    : [constructEmptyResultTallySheet(0)]
+  const formMethods = useForm<IBatchRoundDataEntryFormState>({
+    defaultValues: {
+      resultTallySheets: initialResultTallySheets,
+    },
+  })
+  const {
+    control,
+    formState,
+    getValues,
+    handleSubmit,
+    reset,
+    setValue,
+    watch,
+  } = formMethods
+  const {
+    append: appendSheet,
+    fields: sheets,
+    remove: removeSheet,
+  } = useFieldArray<IBatchResultTallySheet>({
+    control,
+    name: 'resultTallySheets',
+  })
+
+  // Reset state whenever a new batch is selected
+  useEffect(() => {
+    reset({ resultTallySheets: initialResultTallySheets })
+    setSelectedTabId(VOTE_TOTALS_TAB_ID)
+    setIsEditing(false)
+  }, [selectedBatchId])
+
+  // Auto-select first search match
```

Comment:
This felt slightly slow to me when I watched the video. Is it possible to make it a bit more responsive?

Reply:
Ooh I think the root here wasn't the debounce time (250ms) but inconsistent usage of the debounced search query. While the batch selection was being debounced, the batch list filtering was not, resulting in a flicker. I now consistently use the debounced search query: https://github.com/votingworks/arlo/pull/1644/commits/40473164c4164ca8af506f39dd8d6d7e847274cf